src/Idris/Desugar.idr

```diff
@@ -1332,6 +1334,19 @@ mutual
              Overloadable n => pure [IPragma fc [] (\nest, env => setNameFlag fc n Overloadable)]
              Extension e => pure [IPragma fc [] (\nest, env => setExtension e)]
              DefaultTotality tot => pure [IPragma fc [] (\_, _ => setDefaultTotalityOption tot)]
+             ForeignImpl n cs => do
+               cs' <- traverse (desugar AnyExpr ps) cs
+               pure [IPragma fc [] (\nest, env => do
+                      defs <- get Ctxt
+                      calls <- traverse getFnString cs'
+                      [(n',_,gdef)] <- lookupCtxtName n (gamma defs)
+                        | [] => throw (UndefinedName fc n)
+                        | xs => throw (AmbiguousName fc (map fst xs))
+                      let ForeignDef arity xs = gdef.definition
+                        | _ => throw (GenericMsg fc "\{show n} is not a foreign definition")
```

Comment:
Where is this logic coming from?
Is it copy/pasted from the handling of the usual `%foreign` pragma?
In this case, can we instead lift it to the toplevel and share the implementation?

Reply:
That's all new. The `traverse getFnString cs` line was taken from the `%foreign` code (imported from `TTImp.ProcessType`) and `getFnString` is shared. That does the evaluation of the expressions to strings.  The rest is checking to see that there actually is a `%foreign` definition on the other side of the name (and making the name fully qualified in the process).

I could still pull this out to a top level function if we want to streamline the surrounding function.

I do see that I accidentally left in an import (`Core.CompileExpr`) from an earlier attempt.